docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+* **Extensive use of contiguous containers**. Dealing with streams of data is the common case. Contiguous containers offer best performance with O(1) access and perfect cache locality. Can't really say `std::vector` is the best, but it is heavily used.
+* **No runtime Polymorphism**. We don't need polymorphism. Instead, prefer a more true form of message passing through shared variables. Two completely different systems can communicate using queues that act as interfaces for example.
+* **Less Abstractions**. This is not a bad thing, nor does it mean everything is written in assembly. Along with separate data and functions, the computer has separate RAM and CPU (Objects aren't real!). By following a style that more resembles the computer's architecture, less (unnecessary) abstractions are needed to glue together the 'perfect' interface.
+
+Note that this doesn't mean everything has to be ultra-optimized. It's perfectly acceptable to prefer a more convenient option for paths that aren't performance critical, such as operations called only once per frame or once during setup.
+
+## Entity Component System (ECS)
+
+ECS is a solution to representing a world that consists of many 'things' that may have common properties with each other. This is perfect for many games and simulations, and is more versatile and performant than class-based inheritance. Universe and ActiveScene both use ECS through the [EnTT](https://github.com/skypjack/entt/) library.
+
+Confusingly, ECS is not "a system of entities and components" that many mainstream game engines already follow. ECS is better stated as "Entities, Components, and Systems:" three separate concepts that come together to create an elegant architecture.
+
+* **Entity**: A simple unique integer ID linked to the existance of a 'thing'.
+* **Component**: An (optional) property of an entity. Usually a struct stored in a contiguous container that associates it to an Entity.
+* **System**: A function that operates on entities that have a specific set of components. Best described as 'rules' that determine how some components change over time.
+
+Entities are still just integers, but are treated as **compositions** of different types of components. Components represent what an entity **has or is made of**, which determines what operations can be done to it by systems. The components themselves don't describe any behaviour.
+
+As a simple example: any entity with a *Mass component* is affected by the *Gravity system*, any entity with the *Legs component* affected by the *Walk system*. *Mass* or *Legs* don't directly imply any behaviour; it's the world's job to add meaning to them.
+
+Components **never inherit each other**, and all common properties are split into separate components.
```

Comment:
possibly important that it's up to the systems to provide relationships between components, the components themselves don't care.

Which also means that reusing a component name by mistake isn't protected against.